### ISSUE TITLE
Segfault when producing a lot of messages without waiting

### ISSUE THREAD

**Issue by thijsc** [appsignal/rdkafka-ruby#15]
To reproduce produce more than ~ 1000 messages and don't wait for them. I think the Ruby GC cleans up the delivery handles which causes librdkafka to crash when it tries to update them.

**Comment by frsyuki** [comment]
We had the same issue. SEGV happens in producer only under high load.
I suspect that the thread of `rd_kafka_poll` loop is trying to fill DeliveryHandle at `Rdkafka::Bindings::DeliveryCallback` and the memory of the DeliveryHandle is already freed by GC because reference to the DeliveryHandle is lost. Reference is lost when `Rdkafka::Producer#produce` raises an exception, or `DeliveryHandle#wait` finishes (successfully or with an error) but then Kafka sends another message.

**Comment by thijsc** [comment]
Indeed. I haven't found a great way to prevent this yet. We run this under high load, but we add all delivery handles to an array and wait for them. So we don't see this behaviour.

I think there will need to be an internal array that the handles are added to and cleared when the delivery handle finishes.

**Comment by thijsc** [comment]
I've reproduced it in this test:

```ruby
it "should produce message that aren't waited for and not crash" do
  100_000.times do
    producer.produce(
      topic:   "produce_test_topic",
      payload: "payload not waiting",
      key:     "key not waiting"
    )
  end

  # Allow some time for a GC run
  sleep 2
end
```

crashes like this most of the time:

```
rspec spec/rdkafka/producer_spec.rb:163(96098,0x70000e32a000) malloc: *** error for object 0x7fe5c98c84f0: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
[1]    96098 abort      bundle exec rspec spec/rdkafka/producer_spec.rb:163
```

Adding the handle to an array fixes the crash. I'm looking into a way to add this mechanism to the producer internally.

**Comment by mensfeld** [comment]
Even with the fix from the commit I still experienced this issue when running several producers in a spec and not closing them fast enough after the specs were done. I fixed that by defining finalizer hooks to close the client before GC kicks in:

```ruby
ObjectSpace.define_finalizer(self, proc { close })
```

ref: https://github.com/karafka/waterdrop/pull/135/files

**Comment by thijsc** [comment]
It seems like ffi's hooks do not always trigger exactly as I expect. We should maybe switch to this approach altogether.

**Comment by mensfeld** [comment]
I mean for a regular producer that's probably ok as there's usually more than enough time to finish the work before GC kicks in (especially if you close the producer). Rails, Ruby process doing other stuff, Sidekiq etc, but I can imagine this ffi hook not kicking in a problem for thin periodically executable jobs that end with ruby exiting so it might be worth.

**Comment by Adithya-copart** [comment]
@thijsc I think https://github.com/appsignal/rdkafka-ruby/pull/108/commits/f9732b6f0e78604535b30e1488c47dd4ffcac913 is responsible for introducing this in a way.

The handle is no longer a `AutoPointer` object that comes with a finalizer by default. Other changes in https://github.com/appsignal/rdkafka-ruby/pull/108/commits/f9732b6f0e78604535b30e1488c47dd4ffcac913 made it safe to invoke `producer#close` multiple times without a SegFault. So, it could have added a finalizer to invoke `close` safely.

I even added https://github.com/appsignal/rdkafka-ruby/pull/108/commits/6d7b47221fb061b0e97a7eca1a2c1fed98a1f8bf to close all the open sockets but the fix by @mensfeld in https://github.com/appsignal/rdkafka-ruby/pull/115 did not occur to me at the time.

**Comment by mensfeld** [comment]
>  but the fix by @mensfeld in #115 did not occur to me at the time.

@Adithya-copart you mean that after this fix all is good right?

**Comment by Adithya-copart** [comment]
Yeah. The fix looks good to me.

Any producer that is not closed will be closed by the finalizer after GC. If it is already closed by the user, the finalizer will do nothing.
